File: examples/calculator-java-hexagonal-cqrs/src/main/java/com/oopforge/example/calculator/domain/Calculation.java
package com.oopforge.example.calculator.domain;

import java.time.Instant;
import java.util.Objects;

public final class Calculation {

    private final CalculationId id;
    private final double operandA;
    private final Operator operator;
    private final double operandB;
    private final double result;
    private final Instant performedAt;

    private Calculation(CalculationId id, double operandA, Operator operator,
                        double operandB, double result, Instant performedAt) {
        this.id = id;
        this.operandA = operandA;
        this.operator = operator;
        this.operandB = operandB;
        this.result = result;
        this.performedAt = performedAt;
    }

    public static Calculation perform(CalculationId id, double operandA, Operator operator, double operandB) {
        Objects.requireNonNull(id, "id");
        Objects.requireNonNull(operator, "operator");

        double result = operator.apply(operandA, operandB);
        return new Calculation(id, operandA, operator, operandB, result, Instant.now());
    }

    public CalculationId id() {
        return id;
    }

    public double operandA() {
        return operandA;
    }

    public Operator operator() {
        return operator;
    }

    public double operandB() {
        return operandB;
    }

    public double result() {
        return result;
    }

    public Instant performedAt() {
        return performedAt;
    }
}

File: examples/calculator-python-layered/app/calculator/domain/calculation.py
from dataclasses import dataclass
from datetime import UTC, datetime
from enum import Enum
from uuid import UUID, uuid4


class Operator(str, Enum):
    ADD = "add"
    SUBTRACT = "subtract"
    MULTIPLY = "multiply"
    DIVIDE = "divide"

    def apply(self, a: float, b: float) -> float:
        if self is Operator.ADD:
            return a + b
        if self is Operator.SUBTRACT:
            return a - b
        if self is Operator.MULTIPLY:
            return a * b
        if b == 0:
            raise ValueError("division by zero")
        return a / b


@dataclass(frozen=True)
class CalculationId:
    value: UUID

    @staticmethod
    def generate() -> "CalculationId":
        return CalculationId(uuid4())

    @staticmethod
    def parse(raw: str) -> "CalculationId":
        return CalculationId(UUID(raw))


@dataclass(frozen=True)
class Calculation:
    id: CalculationId
    operand_a: float
    operator: Operator
    operand_b: float
    result: float
    performed_at: datetime

    @staticmethod
    def perform(
        calculation_id: CalculationId,
        operand_a: float,
        operator: Operator,
        operand_b: float,
    ) -> "Calculation":
        result = operator.apply(operand_a, operand_b)
        return Calculation._from_result(
            calculation_id, operand_a, operator, operand_b, result
        )

    @staticmethod
    def _from_result(
        calculation_id: CalculationId,
        operand_a: float,
        operator: Operator,
        operand_b: float,
        result: float,
    ) -> "Calculation":
        return Calculation(
            id=calculation_id,
            operand_a=operand_a,
            operator=operator,
            operand_b=operand_b,
            result=result,
            performed_at=datetime.now(UTC),
        )
